nifi-extension-bundles/nifi-standard-services/nifi-oauth2-provider-bundle/nifi-oauth2-provider-service/src/main/java/org/apache/nifi/oauth2/JWTBearerOAuth2AccessTokenProvider.java

```diff
@@ -0,0 +1,677 @@
+        } catch (Exception ex) {
+            String explanation = ex.getMessage();
+            if (ex.getCause() != null) {
+                explanation += " (" + ex.getCause().getMessage() + ")";
+            }
+            builder.outcome(ConfigVerificationResult.Outcome.FAILED).explanation(explanation);
+        }
+        return Arrays.asList(builder.build());
+    }
+
+    @Override
+    public AccessToken getAccessDetails() {
+        if (this.accessDetails == null || isRefreshRequired()) {
+            try {
+                acquireAccessDetails();
+            } catch (Exception e) {
+                throw new AccessTokenRetrievalException("Failed to retrieve access token details", e);
+            }
+        }
+        return accessDetails;
+    }
+
+    private boolean isRefreshRequired() {
+        if (accessDetails.getExpiresIn() > 0) {
+            final Instant expirationRefreshTime = accessDetails.getFetchTime()
+                    .plusSeconds(accessDetails.getExpiresIn())
+                    .minus(refreshWindow);
+
+            return Instant.now().isAfter(expirationRefreshTime);
+        } else {
+            final Instant expirationRefreshTime = accessDetails.getFetchTime()
+                    .plusSeconds(jwtValidity.getSeconds())
+                    .minus(refreshWindow);
+
+            return Instant.now().isAfter(expirationRefreshTime);
+        }
+    }
+
+    private void acquireAccessDetails() throws URISyntaxException, JOSEException {
+        getLogger().debug("New Access Token request started [{}]", tokenEndpoint);
+
+        final Instant now = Instant.now();
+
+        Builder claimsSetBuilder = new JWTClaimsSet.Builder()
+                .expirationTime(Date.from(now.plus(jwtValidity)))
+                .issueTime(Date.from(now));
+
+        if (issuer != null) {
+            claimsSetBuilder.issuer(issuer);
+        }
+
+        if (subject != null) {
+            claimsSetBuilder.subject(subject);
+        }
+
+        if (audience != null) {
+            claimsSetBuilder.audience(Arrays.asList(audience.split(" ")));
+        }
+
+        if (scope != null) {
+            claimsSetBuilder.claim("scope", scope);
+        }
+
+        if (nbf) {
+            claimsSetBuilder.notBeforeTime(Date.from(now));
+        }
+
+        if (jti != null) {
+            claimsSetBuilder.jwtID(jti);
+        }
+
+        customClaims.forEach(claimsSetBuilder::claim);
+
+        JWSHeader.Builder headerBuilder = new JWSHeader.Builder(JWSAlgorithm.parse(algorithmName));
+
+        if (headerType) {
+            headerBuilder.type(JOSEObjectType.JWT);
+        }
+
+        if (kid != null) {
+            headerBuilder.keyID(kid);
+        }
+
+        if (headerX5T) {
+            try {
+                final String alias = keyManager.chooseClientAlias(new String[] {"RSA"}, null, null);
+                if (alias == null) {
+                    getLogger().warn("No key alias found, setting x5t header to null.");
```

Comment:
It is best to avoid ending log messages with a period character. However, shouldn't this throw an exception instead of logging a warning?
```suggestion
                    getLogger().warn("No key alias found, setting x5t header to null");
```

Reply:
So I felt like not setting the field, logging a warning to have a bulletin on the controller service, and still trying to acquire the token was the best approach but happy to consider an exception instead if we feel like it's better to not even try.

Reply:
Given that the property configuration expects the Certificate Thumbprint to be set, I think this should throw an exception as opposed to trying something that seems very unlikely to work.